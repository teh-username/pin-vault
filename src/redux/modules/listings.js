import { combineReducers } from 'redux';

export const ADD_ENTRY = 'modules/listings/ADD_ENTRY';
export const MODIFY_ENTRY = 'modules/listings/MODIFY_ENTRY';
export const DELETE_ENTRY = 'modules/listings/DELETE_ENTRY';

/*
  Sample state:
  {
    entries: {
      2cf05d: {
        name: 'Bank 1',
        code: '1234',
      },
      5d94db: {
        name: 'Bank 2',
        code: '5678',
      },
    },
  }
*/

const initialState = {
  entries: {},
};

export const entries = (state = initialState.entries, action) => {
  switch (action.type) {
    case ADD_ENTRY:
      return {
        ...state,
        [action.id]: {
          code: action.code,
          name: action.name,
        },
      };
    case MODIFY_ENTRY:
      return {
        ...state,
        [action.id]: {
          ...state[action.id],
          code: action.code,
        },
      };
    case DELETE_ENTRY:
      return Object.entries(state)
        .filter(([id, entryDetails]) => action.id !== id)
        .reduce((previousState, [id, entryDetails]) => {
          return {
            ...previousState,
            [id]: entryDetails,
          };
        }, {});
    default:
      return state;
  }
};

export default combineReducers({
  entries,
});

export const addEntry = (id, name, code) => ({
  type: ADD_ENTRY,
  id,
  name,
  code,
});

export const modifyEntry = (id, code) => ({
  type: MODIFY_ENTRY,
  id,
  code,
});

export const deleteEntry = id => ({
  type: DELETE_ENTRY,
  id,
});

export const getEntries = ({ listings: { entries } }) =>
  Object.entries(entries);
export const getEntryDetails = ({ listings: { entries } }, id) => entries[id];
