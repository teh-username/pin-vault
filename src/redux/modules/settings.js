import { combineReducers } from 'redux';

export const TOGGLE_PASSCODE_REQUIREMENT =
  'modules/settings/TOGGLE_PASSCODE_REQUIREMENT';
export const SET_PASSCODE = 'modules/settings/SET_PASSCODE';

/*
  Sample state:
  {
    requirePasscode:  true,
    passcodeHash: '5E884898DA28047151D0E56F8DC6292773603D0D6AABBDD62A11EF721D1542D8',
  }
 */

const initialState = {
  requirePasscode: false,
  passcodeHash: null,
};

const requirePasscode = (state = initialState.requirePasscode, action) => {
  switch (action.type) {
    case TOGGLE_PASSCODE_REQUIREMENT: {
      return !state;
    }
    default:
      return state;
  }
};

const passcodeHash = (state = initialState.passcodeHash, action) => {
  switch (action.type) {
    case SET_PASSCODE:
      return action.passcodeHash;
    default:
      return state;
  }
};

export const togglePasscodeRequirement = () => ({
  type: TOGGLE_PASSCODE_REQUIREMENT,
});

export const setPasscode = passcodeHash => ({
  type: SET_PASSCODE,
  passcodeHash,
});

export default combineReducers({
  requirePasscode,
  passcodeHash,
});

export const getCurrentPasscode = ({ settings: { passcodeHash } }) =>
  passcodeHash;

export const getPasscodeRequirement = ({ settings: { requirePasscode } }) =>
  requirePasscode;
