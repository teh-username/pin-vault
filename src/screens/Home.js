import React from 'react';
import { connect } from 'react-redux';
import { View, StyleSheet } from 'react-native';
import { Icon } from 'react-native-elements';

import EntryList from '../components/EntryList';
import { ADD_ENTRY, getEntries } from '../redux/modules/listings';

export class Home extends React.Component {
  static navigationOptions = ({ navigation }) => ({
    title: 'Entries',
    headerRight: (
      <Icon
        name="menu"
        onPress={() => navigation.navigate('Menu')}
        iconStyle={{ right: 30 }}
      />
    ),
  });

  constructor(props) {
    super(props);
    this.onEntryPress = this.onEntryPress.bind(this);
  }

  onEntryPress(id) {
    this.props.navigation.navigate('Details', { id });
  }

  _renderAddButton(navigation) {
    return (
      <View key="home_add_button" style={styles.icon}>
        <Icon
          raised
          reverse
          name="plus"
          type="font-awesome"
          onPress={() => navigation.navigate('Entry', { action: ADD_ENTRY })}
        />
      </View>
    );
  }

  _renderEntryList(entries) {
    return (
      <EntryList
        key="home_entrylist"
        entries={entries}
        onPress={this.onEntryPress}
      />
    );
  }

  render() {
    const { entries, navigation } = this.props;
    return [this._renderEntryList(entries), this._renderAddButton(navigation)];
  }
}

const styles = StyleSheet.create({
  icon: {
    justifyContent: 'flex-end',
    alignSelf: 'flex-end',
    right: 25,
    bottom: 25,
  },
});

const mapStateToProps = state => ({
  entries: getEntries(state),
});

export default connect(mapStateToProps, null)(Home);
