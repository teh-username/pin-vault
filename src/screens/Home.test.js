import React from 'react';
import { Icon } from 'react-native-elements';

import { Home } from './Home';
import { ADD_ENTRY } from '../redux/modules/listings';

describe('Home screen test', () => {
  it('should render as expected', () => {
    const wrapper = shallow(<Home />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should navigate to Entry screen if the add button is pressed', () => {
    const navMock = {
      navigate: jest.fn(),
    };
    const wrapper = shallow(<Home />);
    const addButton = shallow(
      wrapper.instance()._renderAddButton(navMock)
    ).find(Icon);
    addButton.simulate('press');
    expect(navMock.navigate.mock.calls.length).toBe(1);
    expect(navMock.navigate.mock.calls[0]).toEqual([
      'Entry',
      {
        action: ADD_ENTRY,
      },
    ]);
  });

  it('should navigate to Details screen if an entry is pressed', () => {
    const navMock = {
      navigate: jest.fn(),
    };
    const wrapper = shallow(<Home navigation={navMock} />);
    wrapper.instance().onEntryPress(2);
    expect(navMock.navigate.mock.calls.length).toBe(1);
    expect(navMock.navigate.mock.calls[0]).toEqual(['Details', { id: 2 }]);
  });
});
