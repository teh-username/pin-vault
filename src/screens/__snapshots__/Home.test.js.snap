// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen test should render as expected 1`] = `
Array [
  <EntryList
    key="home_entrylist"
    onPress={[Function]}
  />,
  <View
    key="home_add_button"
    style={
      Object {
        "alignSelf": "flex-end",
        "bottom": 25,
        "justifyContent": "flex-end",
        "right": 25,
      }
    }
  />,
]
`;
